### DISCUSSION TITLE
How to turn a dimmer on/off, and have it remember it's level

### DISCUSSION THREAD

**Discussion by fabnavigator** [zwave-js/node-red-contrib-zwave-js#229] [Q&A]
I have a dimmer light switch that you can manually turn on/off with a quick tap up/down, or you can hold it up/down and it dims up/down.  From node red I would like to set the level, and then turn it off, such that when someone taps it up it goes back to that set level.  Is this possible?

**Comment by marcus-j-davies** [comment]
Hi @fabnavigator,
Does it support `binary switch` (if its listed in the UI, it should do) if so, most devices use that to change the state of the device (in terms of it actually being on or off)

```javascript
let BinarySwitchValueID = {
      "commandClass": 37, /* Binary Switch */
      "endpoint": 0,
      "property": "targetValue",
}
let Message = {
    "payload": {
        "mode": "ValueAPI",
        "node": 46,
        "method": "setValue",
        "params": [BinarySwitchValueID, true | false]
    }
}
return Message
```

**Reply by fabnavigator** [reply]
@marcus-j-davies That code results in:```message: "ZWaveError: Node 30 does not support the Command Class Binary Switch! (ZW0302)"```

**Reply by marcus-j-davies** [reply]
Then the device doesn't support binary.... what's the DB link?

**Comment by fabnavigator** [comment]
I have two similar switches:

https://devices.zwave-js.io/?jumpTo=0x0063:0x4944:0x3130:0.0

https://devices.zwave-js.io/?jumpTo=0x0039:0x4944:0x3130:0.0

**Reply by marcus-j-davies** [reply]
@crxporter,

Could you lend your setup / knowledge in Jasco. In how you restore the previous dim levels when turned on at the switch? (via some device config, or otherwise).

@fabnavigator,

I was hoping to see some config param for the devices that allows to turn on with previous levels - but the devices don't seem to have it.

**Reply by crxporter** [reply]
Hi yep. Well. I'm changing the "on level" config value throughout the day. And using the current value of that when I turn stuff on.

```json
"valueId": {
      "commandClassName": "Configuration",
      "commandClass": 112,
      "endpoint": 0,
      "property": 32,
      "propertyName": "Default Brightness Level"
   }
```

I change this config on all my dimmers like 5 or 6 times daily. I use the current on level value to turn them on.

No idea how to do it in HA but it works great in nodered. You'll need something on the software that knows what the brightness should be when it is restored. 

Anecdotal note: This theme has been ongoing throughout all of my years of home automation on all of the platforms I've used (insteon, zwave, homekit, openhab, nodered, and home assistant have all struggled here in various ways). Clearly it's a complicated thing to sort out...

**Reply by fabnavigator** [reply]
@crxporter What type of dimmers do you have? My dimmers only hve these configuration properties:
```
Dim Rate Steps (Z-Wave Command)
Dim Rate Timing (Z-Wave)
Dim Rate Steps (Manual)
Dim Rate Timing (Manual)
Dim Rate Steps (All-On/All-Off)
Dim Rate Timing (All-On/All-Off)
```

**Reply by crxporter** [reply]
Ah- I just clicked your links, it looks like yours are older than mine. I have various (Honeywell, ultra pro, ge... whatever's discounted the day of purchase) purchased over the last 2 years or so. 

All of my database files look like these:
https://devices.zwave-js.io/?jumpTo=0x0063:0x4944:0x3339:0.0
https://devices.zwave-js.io/?jumpTo=0x0039:0x4944:0x3235:0.0

**Reply by marcus-j-davies** [reply]
@fabnavigator,

The way to achieve it, without the config being available on the device, and how I believe  HA are likely doing it.

Just store the dim level, when ever it changes (the device should send an update if changed at the device). Then when you turn it on remotely (I.e node red) just turn it on setting the multilevel (or basic) level, that was previously saved, during the last update.

Basic or Multilevel set at 0 is basically off, anything else is on at the set level.

Just store the set level when ever it changes, and use it for the on value.

Again, this is likely what HA is doing.

**Reply by fabnavigator** [reply]
@marcus-j-davies 

Adjusting the brightness from NR isn't an issue. The goal is to use NR to set the level to say 50%, and then turn off the light. This is what I can do from HA. Then when someone manually turns on the light it goes to 50%.

@crxporter can set the default brightness level w/o even turning on the light. That is even better.

I'm going to move one of my dimmers back to HA and see if I can see what commands/configuration settings are sent to the device.

If I can't fugure it out. I guess I'll buy some new dimmers that have a default brightness level configuration parameter.

**Reply by marcus-j-davies** [reply]
The device supports Scene Actuator Configuration CC.
So maybe its using that 🤷‍♂️ 

Set the scene ID, and its level (I'm going to assume scene ID 1 is the unit being turned on - but its device specific)

```javascript
let Message = {
    payload: {
        mode: "CCAPI",
        node: 45,
        cc: 0x2c  /*Scene Actuator Conf */,
        method: "set",
        params: [1, undefined, 50] /* Scene ID, Transition Time (example: "0.5s"), Level */
    }
}
return Message
```
Reference.
[Scene Actuator Conf ](https://zwave-js.github.io/node-zwave-js/#/api/CCs/SceneActuatorConfiguration?id=set)

i.e the device my trigger an internal scene ID of 1 when turned on.

**Reply by fabnavigator** [reply]
@marcus-j-davies 

Thank you. I'll give it a try, but it will be a couple of days, I'm going out of town.

**Reply by crxporter** [reply]
@fabnavigator regarding this:

> can set the default brightness level w/o even turning on the light. That is even better.

My dimmers "on level" is for turning on *at the switch*. That value has nothing to do with my automations. I simply remember the value in my home automation system so I can use it when needed.

**Comment by fabnavigator** [comment]
Here is what the dimmer control looks like in Home Assistant. It works the way I need it to. You can set the brightness, but also have an on/off switch that doesn't adjust the set brightness level.

![Dimmer](https://user-images.githubusercontent.com/16625904/177043079-2992f85a-8d6c-44f3-b5d8-464ab39a7d7b.png)

**Comment by marcus-j-davies** [comment]
I cant comment on HA.

And how are you turning it off in NR and back on?
if setting its Multilevel to 0 - just store the Multilevel every time it changes.
you should also get the value - when dimmed at the switch

Then once being turned on in NR - just use the last captured Multi Level ON value.

if the switch does not support Binary - that maybe how HA is doing it
